packages/xrpc-server/src/server.ts

```diff
@@ -200,6 +240,91 @@ export class Server {
       }
     }
   }
+
+  protected async addSubscription(
+    nsid: string,
+    def: LexXrpcSubscription,
+    config: XRPCStreamHandlerConfig,
+  ) {
+    const assertValidXrpcParams = (params: unknown) =>
+      this.lex.assertValidXrpcParams(nsid, params)
+    const resolveLexUri = (ref: string) => this.lex.resolveLexUri(nsid, ref)
+    this.subscriptions.set(
+      nsid,
+      new XrpcStreamServer({
+        noServer: true,
+        handler: async function* (req) {
+          try {
+            const url = new URL(req.url || '', 'http://x')
+            // authenticate request
+            const auth = await config.auth?.({ req })
+            if (isHandlerError(auth)) {
+              throw XRPCError.fromError(auth)
+            }
+            // validate request
+            const params = decodeQueryParams(
+              def,
+              getQueryParams(url.searchParams),
+            )
+            try {
+              assertValidXrpcParams(params)
+            } catch (e) {
+              throw new InvalidRequestError(String(e))
+            }
+            // stream
+            const items = config.handler({ req, params, auth })
+            for await (const item of items) {
+              if (item instanceof Frame) {
+                yield item
+              } else if (
+                typeof item?.['$type'] === 'string' &&
+                def.output?.codes
+              ) {
+                const typeUri = resolveLexUri(item['$type'])
+                if (def.output.codes[typeUri] !== undefined) {
+                  const code = def.output.codes[typeUri]
+                  const clone = { ...item }
+                  delete clone['$type']
+                  yield new DataFrame({ type: code, body: clone })
+                } else {
+                  yield new DataFrame({ body: item })
+                }
+              } else {
+                yield new DataFrame({ body: item })
+              }
+            }
+          } catch (err) {
+            const xrpcErrPayload = XRPCError.fromError(err).payload
```

Comment:
Looks like we might be leaking internal server error messages here?

Reply:
I think we should be all set on that, it should be handled inside the `payload` getter.